### PR TITLE
Shrink `ast::Attribute`.

### PR DESCRIPTION
**Pull request by nnethercote**
r? @ghost

### CODE REVIEW

**Comment by nnethercote** [comment]
Surprisingly good results on `doc` profiles, especially for max-rss. I don't know why `doc` profiles would be more sensitive to AST changes than other profilers. `doc` builds often use a lot more memory, too, e.g. max-rss (last two columns show the actual max-rss numbers):

Benchmark | Profile | Scenario | % Change | Significance Factor? | #99464 | #100441
-- | -- | -- | -- | -- | -- | --
helloworld | doc | full | -5.80% | 5.59x | 244,436 | 230,252
helloworld | check | full | 0.03% | 0.02x | 84,708 | 84,732

cc @jyn514

**Comment by jyn514** [comment]
> I don't know why `doc` profiles would be more sensitive to AST changes than other profilers. `doc` builds often use a lot more memory, too, e.g. max-rss (last two columns show the actual max-rss numbers):

that doesn't surprise me too much - everything in https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/clean/types/index.html is basically a wrapper around `rustc_ast` or `rustc_hir`, and we duplicate a lot of stuff in DocContext that's already in TyCtxt.

Glad to see this improve the performance so much!

**Comment by jyn514** [comment]
I'm not a good reviewer for this code.

r? @petrochenkov maybe

File: compiler/rustc_ast/src/ast.rs [hunk 1]
```diff
@@ -2547,10 +2547,17 @@ pub struct Attribute {
     pub span: Span,
 }
 
+/// Metadata associated with an item.
+#[derive(Clone, Encodable, Decodable, Debug)]
+pub struct Normal {
```
Comment (petrochenkov) [inline compiler/rustc_ast/src/ast.rs:2552]:
```suggestion
pub struct NormalAttr {
```

**Comment by petrochenkov** [comment]
Seems like a win on average.
r=me after renaming `ast::Normal`.
@rustbot author

**Comment by nnethercote** [comment]
Fun fact: this change almost halves the amount of memcpy'ing done when running `rustdoc` on `helloworld`:

Old:
```
Total:     297,675,803 bytes (100%, 202,140.01/Minstr) in 869,856 blocks (100%, 590.69/Minstr), avg size 342.21 bytes
```
New:
```
Total:     180,117,453 bytes (100%, 128,733.54/Minstr) in 438,980 blocks (100%, 313.75/Minstr), avg size 410.31 bytes
```

**Comment by nnethercote** [comment]
I renamed `Normal` as `NormalAttr`.

@bors r=petrochenkov

**Comment by nnethercote** [comment]
Third time lucky:

@bors r=petrochenkov

**Comment by klensy** [comment]
There some ast trait that implemented for some structs with tokens. Should it be implemented for `NormalAttr`?
https://github.com/rust-lang/rust/blob/8556e6620e4866526b3cea767ad8c20ae877a569/compiler/rustc_ast/src/ast_traits.rs#L131-L162

**Comment by nnethercote** [comment]
> There some ast trait that implemented for some structs with tokens. Should it be implemented for `NormalAttr`?

It is covered here:
https://github.com/rust-lang/rust/blob/8556e6620e4866526b3cea767ad8c20ae877a569/compiler/rustc_ast/src/ast_traits.rs#L212-L229

**Pull request merged**